File: frontend/lib/api/standards.ts
import { useQuery, useMutation, UseQueryResult, UseMutationResult } from 'react-query'
import { apiClient } from './client'

export interface StandardFramework {
  id: string
  name: string
  authority: string
  jurisdiction: string
  version: string
  description: string
  created_at: string
  updated_at: string
}

export interface Standard {
  id: string
  tenant_id: string
  framework_id: string
  parent_id?: string
  code: string
  description: string
  grade?: string
  subject?: string
  domain?: string
  strand?: string
  version: string
  created_at: string
  updated_at: string
}

export interface StandardHierarchy {
  framework: StandardFramework
  standards: Standard[]
  hierarchy: {
    id: string
    code: string
    label: string
    children: any[]
  }[]
}

// Fetch all frameworks
export const useFrameworks = (): UseQueryResult<StandardFramework[], Error> => {
  return useQuery(
    ['frameworks'],
    async () => {
      const response = await apiClient.get<StandardFramework[]>('/api/v1/standards/frameworks')
      return response.data
    },
    {
      staleTime: 5 * 60 * 1000, // 5 minutes
      cacheTime: 30 * 60 * 1000, // 30 minutes
    }
  )
}

// Fetch standards in a framework
export const useStandards = (frameworkId: string): UseQueryResult<Standard[], Error> => {
  return useQuery(
    ['standards', frameworkId],
    async () => {
      const response = await apiClient.get<Standard[]>(
        `/api/v1/standards/frameworks/${frameworkId}/standards`
      )
      return response.data
    },
    {
      enabled: !!frameworkId,
      staleTime: 5 * 60 * 1000,
    }
  )
}

// Fetch standard hierarchy
export const useStandardHierarchy = (
  frameworkId: string
): UseQueryResult<StandardHierarchy, Error> => {
  return useQuery(
    ['standard-hierarchy', frameworkId],
    async () => {
      const response = await apiClient.get<StandardHierarchy>(
        `/api/v1/standards/frameworks/${frameworkId}/hierarchy`
      )
      return response.data
    },
    {
      enabled: !!frameworkId,
      staleTime: 10 * 60 * 1000,
    }
  )
}

// Fetch single standard detail
export const useStandard = (standardId: string): UseQueryResult<Standard, Error> => {
  return useQuery(
    ['standard', standardId],
    async () => {
      const response = await apiClient.get<Standard>(`/api/v1/standards/${standardId}`)
      return response.data
    },
    {
      enabled: !!standardId,
      staleTime: 5 * 60 * 1000,
    }
  )
}

// Search standards
export const useSearchStandards = (
  query: string,
  frameworkId?: string
): UseQueryResult<Standard[], Error> => {
  return useQuery(
    ['standards-search', query, frameworkId],
    async () => {
      const params = new URLSearchParams()
      if (query) params.append('q', query)
      if (frameworkId) params.append('framework_id', frameworkId)

      const response = await apiClient.get<Standard[]>(
        `/api/v1/standards/search?${params.toString()}`
      )
      return response.data
    },
    {
      enabled: !!query,
      staleTime: 1 * 60 * 1000, // 1 minute for search
    }
  )
}

// Get standards by grade and subject
export const useStandardsByGradeSubject = (
  frameworkId: string,
  grade?: string,
  subject?: string
): UseQueryResult<Standard[], Error> => {
  return useQuery(
    ['standards-grade-subject', frameworkId, grade, subject],
    async () => {
      const params = new URLSearchParams()
      if (grade) params.append('grade', grade)
      if (subject) params.append('subject', subject)

      const response = await apiClient.get<Standard[]>(
        `/api/v1/standards/frameworks/${frameworkId}/by-grade-subject?${params.toString()}`
      )
      return response.data
    },
    {
      enabled: !!frameworkId && (!!grade || !!subject),
      staleTime: 5 * 60 * 1000,
    }
  )
}

// Mutation: Create standard framework
export const useCreateFramework = (): UseMutationResult<
  StandardFramework,
  Error,
  Omit<StandardFramework, 'id' | 'created_at' | 'updated_at'>,
  unknown
> => {
  return useMutation(async (data) => {
    const response = await apiClient.post<StandardFramework>(
      '/api/v1/standards/frameworks',
      data
    )
    return response.data
  })
}

// Mutation: Create standard
export const useCreateStandard = (): UseMutationResult<
  Standard,
  Error,
  Omit<Standard, 'id' | 'tenant_id' | 'created_at' | 'updated_at'>,
  unknown
> => {
  return useMutation(async (data) => {
    const response = await apiClient.post<Standard>('/api/v1/standards', data)
    return response.data
  })
}

File: backend/api/standards.py
"""Standards API endpoints - browse frameworks and standards."""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from database.db import get_db
from database.models import StandardFramework, Standard
from pydantic import BaseModel

router = APIRouter(prefix="/v1/standards", tags=["standards"])


# Response models
class StandardResponse(BaseModel):
    id: str
    code: str
    description: str
    grade: Optional[str] = None
    subject: Optional[str] = None
    domain: Optional[str] = None
    strand: Optional[str] = None
    version: Optional[str] = None

    class Config:
        from_attributes = True


class StandardFrameworkResponse(BaseModel):
    id: str
    name: str
    authority: Optional[str] = None
    jurisdiction: Optional[str] = None
    version: Optional[str] = None
    description: Optional[str] = None

    class Config:
        from_attributes = True


class StandardHierarchyItem(BaseModel):
    id: str
    code: str
    label: str
    children: List["StandardHierarchyItem"] = []


StandardHierarchyItem.model_rebuild()


class StandardHierarchyResponse(BaseModel):
    framework: StandardFrameworkResponse
    standards: List[StandardResponse]
    hierarchy: List[StandardHierarchyItem]


# ==================== Framework Endpoints ====================

@router.get("/frameworks", response_model=List[StandardFrameworkResponse])
async def list_frameworks(db: Session = Depends(get_db)):
    """List all standard frameworks."""
    frameworks = db.query(StandardFramework).all()
    return frameworks


@router.get("/frameworks/{framework_id}", response_model=StandardFrameworkResponse)
async def get_framework(framework_id: str, db: Session = Depends(get_db)):
    """Get a specific standard framework."""
    framework = db.query(StandardFramework).filter(
        StandardFramework.id == framework_id
    ).first()

    if not framework:
        raise HTTPException(status_code=404, detail="Framework not found")

    return framework


# ==================== Standards Endpoints ====================

@router.get("/frameworks/{framework_id}/standards", response_model=List[StandardResponse])
async def get_framework_standards(
    framework_id: str,
    grade: Optional[str] = Query(None),
    subject: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Get all standards in a framework, with optional grade/subject filtering."""
    # Verify framework exists
    framework = db.query(StandardFramework).filter(
        StandardFramework.id == framework_id
    ).first()

    if not framework:
        raise HTTPException(status_code=404, detail="Framework not found")

    # Query standards
    query = db.query(Standard).filter(Standard.framework_id == framework_id)

    if grade:
        query = query.filter(Standard.grade == grade)
    if subject:
        query = query.filter(Standard.subject == subject)

    standards = query.order_by(Standard.code).all()
    return standards


@router.get("/{standard_id}", response_model=StandardResponse)
async def get_standard(standard_id: str, db: Session = Depends(get_db)):
    """Get a specific standard."""
    standard = db.query(Standard).filter(Standard.id == standard_id).first()

    if not standard:
        raise HTTPException(status_code=404, detail="Standard not found")

    return standard


@router.get("/frameworks/{framework_id}/by-grade-subject", response_model=List[StandardResponse])
async def get_standards_by_grade_subject(
    framework_id: str,
    grade: Optional[str] = Query(None),
    subject: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Get standards filtered by grade and/or subject."""
    # Verify framework exists
    framework = db.query(StandardFramework).filter(
        StandardFramework.id == framework_id
    ).first()

    if not framework:
        raise HTTPException(status_code=404, detail="Framework not found")

    # Query standards
    query = db.query(Standard).filter(Standard.framework_id == framework_id)

    if grade:
        query = query.filter(Standard.grade == grade)
    if subject:
        query = query.filter(Standard.subject == subject)

    standards = query.order_by(Standard.code).all()
    return standards


@router.get("/search", response_model=List[StandardResponse])
async def search_standards(
    q: Optional[str] = Query(None),
    framework_id: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Search standards by code or description."""
    query = db.query(Standard)

    if framework_id:
        query = query.filter(Standard.framework_id == framework_id)

    if q:
        # Search in code and description (case-insensitive)
        search_term = f"%{q.lower()}%"
        query = query.filter(
            (Standard.code.ilike(search_term)) |
            (Standard.description.ilike(search_term))
        )

    standards = query.order_by(Standard.code).limit(50).all()
    return standards


# ==================== Hierarchy Endpoints ====================

@router.get("/frameworks/{framework_id}/hierarchy", response_model=StandardHierarchyResponse)
async def get_standard_hierarchy(framework_id: str, db: Session = Depends(get_db)):
    """Get standards organized in a hierarchical structure by grade."""
    # Fetch framework
    framework = db.query(StandardFramework).filter(
        StandardFramework.id == framework_id
    ).first()

    if not framework:
        raise HTTPException(status_code=404, detail="Framework not found")

    # Fetch all standards
    standards = db.query(Standard).filter(
        Standard.framework_id == framework_id
    ).order_by(Standard.code).all()

    # Group by grade
    standards_by_grade = {}
    for std in standards:
        grade = std.grade or "Ungrouped"
        if grade not in standards_by_grade:
            standards_by_grade[grade] = []
        standards_by_grade[grade].append(std)

    # Build hierarchy (group by grade and domain)
    grade_order = ["K", "1", "2", "3", "4", "5", "6", "7", "8", "9", "9-10", "10", "11", "11-12", "12", "Ungrouped"]

    hierarchy_items = []
    for grade in grade_order:
        if grade not in standards_by_grade:
            continue

        grade_standards = standards_by_grade[grade]

        # Group by domain within grade
        domains = {}
        for std in grade_standards:
            domain = std.domain or "General"
            if domain not in domains:
                domains[domain] = []
            domains[domain].append(std)

        # Build grade item with domain children
        domain_items = []
        for domain_name in sorted(domains.keys()):
            domain_stds = domains[domain_name]
            standard_items = [
                StandardHierarchyItem(
                    id=std.id,
                    code=std.code,
                    label=f"{std.code}: {std.description[:60]}..."
                )
                for std in domain_stds
            ]
            domain_items.append(
                StandardHierarchyItem(
                    id=f"domain_{grade}_{domain_name}",
                    code=domain_name,
                    label=f"{domain_name} ({len(domain_stds)})",
                    children=standard_items
                )
            )

        grade_item = StandardHierarchyItem(
            id=f"grade_{grade}",
            code=grade,
            label=f"Grade {grade} ({len(grade_standards)})",
            children=domain_items
        )
        hierarchy_items.append(grade_item)

    return StandardHierarchyResponse(
        framework=StandardFrameworkResponse.model_validate(framework),
        standards=[StandardResponse.model_validate(s) for s in standards],
        hierarchy=hierarchy_items
    )
